internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// PriorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type PriorityQueue struct {
+	queue []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+
+	m sync.Mutex
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *PriorityQueue) Dequeue() (wrp.Message, bool) {
+	defer pq.m.Unlock()
```

Comment:
Usually this is done in the other order in go.  (`Lock()`, then defer `Unlock()`)